lib/puppet/resource.rb

```diff
@@ -445,7 +445,7 @@ def to_hierayaml
 
     attributes = attr.collect { |k|
       v = parameters[k]
-      "    %-#{attr_max}s: %s\n" % [k, Puppet::Parameter.format_value_for_display(v)]
+      "    %-#{attr_max}s: %s\n" % [k, v.inspect]
```

Comment:
This doesn't look right to me. `format_value_for_display` is supposed to handle serialization from ruby types to safe puppet data types: https://puppet.com/docs/puppet/latest/lang_data_abstract.html#the-data-data-type. For example stringifying symbols to strings, etc.

It looks like the pops string converter doesn't recognize escaped single quotes in a single or double quoted string:

```
irb(main):003:0> Puppet::Parameter.format_value_for_display('a\'b')
=> "'a\\'b'"
irb(main):004:0> Puppet::Parameter.format_value_for_display("a\'b")
=> "'a\\'b'"
irb(main):005:0> Puppet::Parameter.format_value_for_display("a'b")
=> "'a\\'b'"
```

I would expect `"a'b"` to be returned in all cases. Thoughts @hlindberg?

Reply:
@joshcooper Thanks for the brain puzzle...

The `format_value_for_display` result is a puppet language string that  when printed out and then copy/pasted into puppet source would yield the original string. So... if you do a `puts()` of that inspected value you get: `'a\'b'` - which if that is placed in puppet source results in the string `a'b` which is what you expected the result to be. So, it works as intended.

The `format_value_for_display` is hower not the right choice for turning data into YAML (it is puppet source output, not YAML), nor `inspect` (also not YAML, since it is Ruby source (kind of, but not guaranteed)). Should probably use the methods for serializing to be 100% correct - there could for example be rich data in the result.

Reply:
could we call `YAML.dump(v)` ?

Reply:
Or is there something in `Pops` that exists we could use instead?

Reply:
@jtappa The `YAML.dump()` way is probably a safer way to go, but the `to_hierayaml` function does not output a valid YAML document (no leading `---`, 2 extra indentation spaces).

In my understanding, this is expected because the calling function prepends the type at the beginning.  Maybe a more invasive but less error prone refactoring would be to return a YAML agnostic data structure, and in the calling function aggregate these data structures and feed `YAML.dump()` with it?

Reply:
This refactoring is actually trivial :smile: With some extra work (so that symbols and dates are converted to string, I obtain similar YAML documents for the valid cases:

```sh-session
romain@zappy ~/Projects/puppet % puppet resource file /etc --to_yaml
file:
  /etc:
    ensure: 'directory'
    ctime : '2019-08-23 19:47:52 -1000'
    group : 0
    mode  : '0755'
    mtime : '2019-08-23 19:47:52 -1000'
    owner : 0
    type  : 'directory'
romain@zappy ~/Projects/puppet % bundle exec bin/puppet resource file /etc --to_yaml
---
file:
  "/etc":
    ensure: directory
    ctime: '2019-08-23 19:47:52 -1000'
    group: 0
    mode: '0755'
    mtime: '2019-08-23 19:47:52 -1000'
    owner: 0
    type: directory
```

Key differences:
* We now have a `---` preamble emitted;
* Strings are only quoted when needed (e.g. single quote for dates, double quotes for path starting with a `/`, no quotes for single words…  all this is done by `YAML.dump`);
* `:` are not vertically aligned anymore (cosmetic only);
* Resources are not separated by a blank line anymore (cosmetic only).

All these changes should be innocent as far as YAML is concerned.  Let's amend my last commit to check this!